kernel/events.c

```diff
@@ -269,7 +269,18 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 	SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_event, wait, event, events,
 					   options, timeout);
 
-	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
+	int status = z_pend_curr(&event->lock, key, &event->wait_q, timeout);
+#if CONFIG_MP_NUM_CPUS > 1
+	/*
+	 * On SMP systems, threads in k_event_post_internal()'s processing
+	 * loop may timeout before the list finishes processing. The following
+	 * spin lock/unlock pair prevents those threads from executing further
+	 * until that loop is finished processing.
```

Comment:
I had trouble when this came by the first time.  The effect of the lock/unlock is to prevent **this thread** from continuing until the thread that holds the lock (if one exists) releases it.  I don't think that's consistent with the comment here, which seems to imply that this is going to block other threads, which it can't,... except for the few cycles it holds the lock, during which it's not taking any externally-visible action anyway, so it might as well not be happening from the perspective of other threads.  So... why must **this** thread block, and what might happen if it doesn't?

Can we get a sequence diagram of the synchronization here?  It still feels off to me.

Reply:
> The effect of the lock/unlock is to prevent this thread from continuing until the thread that holds the lock (if one exists) releases it.

And to be clear: ...which is a very different state than "no other thread now holds the lock" or "no other thread was able to take the lock and release it a second/third/etc... time"!

Reply:
@andyross So k_event_wait_internal() waits for the events to be posted. Now let's say core 1 is executing the k_event_post_internal() do/while loop to unpend the threads while walking the linked list.  Now, if a thread on core 2 that was waiting for an event gets timed out while core 1 was in the mid of executing the do/while loop, then thread 2 would be removed from the waitq and will be unpended. Following that sequence thread 2 now would continue execution while we are still walking the linked list in the do/while loop (k_event_post_internal). This could lead to corruption of thread->next_event_link and thread->events. We want to avoid this behavior in SMP systems, hence I say "prevents those threads from executing". Let me know if is confusing from the reader's perspective.

Tried to answer your question. Let me know if it doesn't make sense.

Reply:
> Following that sequence thread 2 now would continue execution while we are still walking the linked list in the do/while loop

But how does this code stop thread 2?  Even if it starts spinning on the lock taken here, it will exit that spin loop instantaneously when we release the lock.  You "stop" a thread from progressing by holding a lock, not just taking it.

Now, there **are** synchronization paradigms that can work purely with ordered events.  I'm not saying this can't possibly be fixing a bug.  I'm just saying that if we want to commit a novel lockless algorithm we need to do a lot more work to explain things.

Mostly I just don't buy that this is fixing anything.   Can't we rework this so that it uses traditional synchronization paradigms?

Reply:
I think I see the core of the misunderstanding, and it stems from description in the comment--the spin lock barrier is meant to prevent the current thread from executing.

In the following scenario, thread T1 on core 1 is processing the list of threads to wake up in k_event_post_internal(). Thread T2 is one of the threads in that list. The intent of the spin lock barrier in k_event_wait_internal() is to prevent thread T2 from executing further (on a different core) if it had timed out while T1 is processing that list. As T1 is holding the spin lock, T2 will wait until that spin lock has been released before (safely) continuing.

What are the worst case scenarios should T2 progress before T1 finishes? T2 could wait on another event object, and then when T1 gets around to processing the T2 node in its list, T1 may change the set of events that T2 is now waiting on. Or, T2 could now be part of a different chain of threads being handled by T3, and when T2 is processed by T1, T1 will continue processing T3's chain.

Such scenarios can occur on SMP because the timeout queue and the event's wait queue are protected by spin locks other than the event's spin lock. It does not occur on UP because there are no other cores in play, and the event's spin lock degenerates into an interrupt lock which is sufficient for that case.

Is that spin lock barrier elegant? No. Does it work? I think it does.

Reply:
I'm still confused.  Is T1 the thread taking the lock here or the one running on a different cpu?  Could you write this up in event list form (i.e. "Thread X on core A reaches line M; Thread Y on core B is at line N; ..."), explain the broken part in detail (i.e. "X may read Q in one of two different states depending on the race with Y") and then add that to the comment?

But to repeat: the only (!) reliable effect of this code here is to cause k_event_wait_internal() to wait until someone else releases the lock.  So the explanation of the fix should have the form "Once k_spin_lock() returns, we are guaranteed that <insert safe state here>"

Reply:
1. T2 waits on event object A. Events for which it is waiting are identified in T2’s events field.
2. T1 (on core 1) takes the event object A’s spin lock.
3. T1 creates a list of threads to wake linked together via each thread’s next_event_link field.
4. Core 2 processes a timer interrupt. This removes T2 from event object A’s wait queue.
5. T2 wakes and runs on core 2. It is now in an unpredictable state because it is running.
6. Meanwhile on core 1, T1 finally gets to processing T2 in its list of threads to process (created in step 3).
Here comes the meat of the problem ...
7. T1 (which still has event object A’s spin lock) currently updates and references T2’s fields. However, this is unsafe to do so because T2 is not in a known state. Since T2 previously executed, it could be waiting on a different event object (making it more unsafe for T1 to modify/use T2’s events or next_event_link fields).

We can patch this hole by tweaking step 5 above to become ....
5. T2 wakes and runs on core 2. However, before it can do anything of significance, we have it busy wait on event object A’s spin lock.
6. Meanwhile on core 1, T1 finally gets to processing T2 in its list of threads to process (created in step 3).
7. T1 can now safely use T2’s events and next_event_link fields as T1 has the spin lock and T2 is waiting for it.
8. When T1 finishes processing the list it releases the spin lock.
9. T2 is now able to obtain the spin lock and continue.
10. T2 immediately releases it because now it is the only thread that might be updating its events and next_event_link fields.